Split out daikin IR protocol and Daikin climate component

wjtje opened the pull request:
# What does this implement/fix?

This MR moved the raw Daikin IR data to usable bytes to class under the `remote_base`, this has the following advantages:
- Able to send custom data using actions
- Able to see what data is received using `remote_receiver` dump
- Remove responsibilities from the Daikin ir climate component

<!-- Quick description and explanation of changes -->
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4659

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
remote_receiver:
  - id: ir_receiver
    pin:
      number: GPIO4
      inverted: True
    dump:
      - daikin

remote_transmitter:
  - id: ir_transmitter
    pin:
      number: GPIO15
    carrier_duty_percent: 50%

climate:
  - platform: daikin
    name: "Daikin AC"
    receiver_id: ir_receiver
    transmitter_id: ir_transmitter
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

wjtje committed:
Spit out daikin IR protocol

wjtje committed:
Added tests for daikin protocol

wjtje committed:
Change loop to a range based loop

DjordjeMandic commented on esphome/components/remote_base/daikin_protocol.h:
```diff
@@ -0,0 +1,42 @@
+ public:
+  TEMPLATABLE_VALUE(std::vector<uint8_t>, data);
+
+  void set_data(const std::vector<uint8_t> &data) { data_ = data; }
```
```suggestion
  void set_data(const std::vector<uint8_t> &data) { this->data_ = data; }
```

DjordjeMandic commented on esphome/components/remote_base/daikin_protocol.cpp:
```diff
@@ -0,0 +1,110 @@
+
+void DaikinProtocol::encode(RemoteTransmitData *dst, const DaikinData &data) {
+  dst->set_carrier_frequency(IR_FREQUENCY);
+
```
Reserve enough space before adding items into dst to prevent frequent memory realloc.

DjordjeMandic commented on esphome/components/remote_base/daikin_protocol.cpp:
```diff
@@ -0,0 +1,110 @@
+
+    if (i < sizeof(HEADER)) {
+      if (*it != HEADER[i]) {
+        ESP_LOGD(TAG, "Invalid header byte (%hhu): received %02X, expected %02X", i, *it, HEADER[i]);
```
```suggestion
        ESP_LOGD(TAG, "Invalid header byte (%" PRIu8 "): received %02X, expected %02X", i, *it, HEADER[i]);
```

```suggestion
        ESP_LOGD(TAG, "Invalid header byte (%u): received %02X, expected %02X", i, *it, HEADER[i]);
```

Size prefix hh is used only for scanf Afaik, in printf, before printing everything gets cast to largest integer. (it's best to check docs for each platform or just use format macro specifier to make it compatible across multiple platforms)

DjordjeMandic commented on esphome/components/remote_base/daikin_protocol.cpp:
You can use item to combine mark and space.

wjtje committed:
Improved the daikin protocol
- Reserve enough space for the remote transmit data
- Now using cinttypes instead of hard coded values
